src/ai/backend/manager/idle.py

```diff
@@ -307,24 +305,148 @@ class UtilizationIdleChecker(BaseIdleChecker):
+        await self._redis.set(
+            f"session.{session_id}.last_execution", "0", exist=self._redis.SET_IF_EXIST
+        )
+
+    async def _update_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"ExecutionIdleChecker._update_timeout({session_id})")
+        t = await self._redis.time()
+        await self._redis.set(
+            f"session.{session_id}.last_execution",
+            f"{t:.06f}",
+            expire=float(sum(self.cpu_util_series) / len(self.cpu_util_series))
+        )
+
+    async def _session_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: SessionStartedEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _execution_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionStartedEvent,
+    ) -> None:
+        await self._disable_timeout(event.session_id)
+
+    async def _execution_exited_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionFinishedEvent | ExecutionTimeoutEvent | ExecutionCancelledEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _do_idle_check(
+        self,
+        context: None,
+        source: AgentId,
+        event: DoIdleCheckEvent,
+    ) -> None:
+
+        try:
+            return await super()._do_idle_check(context, source, event)
+        finally:
+            pass
+
+    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
+        session_id = session["id"]
+        active_streams = await self._redis.zcount(
+            f"session.{session_id}.active_app_connections"
+        )
+        if active_streams is not None and active_streams > 0:
+            return True
+
+        redis = await aioredis.create_redis("redis://127.0.0.1:8111/")
+        redis_out = await redis.get(str(session["session_id"]), encoding=None)
+        try:
+            live_stat = msgpack.unpackb(redis_out)
+        except Exception:
+            return True
+
+        cpu_util_pct = float(live_stat["cpu_util"]["pct"])
+        mem_util_pct = float(live_stat["mem"]["pct"])
+        cuda_mem_util_pct = float(live_stat["cuda_mem"]["pct"])
+
+        interval = self.timer.interval
+
+        if "m" in self.window:
+            window = int(self.window.replace("m", ""))
+            window_size = (window * 60) / interval
+        elif "h" in self.window:
+            window = int(self.window.replace("h", ""))
+            window_size = (window * 60 * 60) / interval
+        else:
+            raise Exception("Undefined unit of window interval")
+
+        if len(self.cpu_util_series) < window_size:
+            self.cpu_util_series.append(float(cpu_util_pct))
+            self.mem_util_series.append(float(mem_util_pct))
+            self.cuda_mem_util_series.append(float(cuda_mem_util_pct))
+            return True
+        else:
```

Comment:
Don't we need to append `*_util_pct` to `self.*_util_series` in `else` block as well? I can see `pop`ing oldest elements out, but not appending new values.

Reply:
Overall, here is implemented the sliding moving window with step 1. After window fills up until defined window size. It will be always popping the oldest value and adding the newest value. So that window moves by 1 element.  And in that window we continuously measure the average and if the threshold is reached. There is no loop, but session_checker()  repetitively called by _do_idle_checker(), so the adding and removing elements implemented in if/else block.

The execution of the code is following:
If a number of elements are less than the defined window size then append a new element from Redis and return True.
Since returned value True doesn't terminate the session, the session_checker() is called again by _do_idle_check().
It checks the window size and sees that it again has fewer elements than the defined window size. So it appends and session_checker is called until a number of elements become equal to the defined window size.

At that point, in Else condition. The average is measured for all values in a window. Now we don't need the oldest element anymore, so it gets popped out. Then we check if the threshold condition is met. If a threshold is not met return True value. After that session_checker will be triggered again by _do_idle_check() and since a number of elements in the window are less than the defined window size, then the newest value will be appended and then Else condition will be triggered, and the threshold will be again checked.

Reply:
Yes, I understand what you mean, and the logic will work. But, I wonder is there a reason not to append new data in the `else` block and not going back to the `if` block any more once the window size is reached.

Also, the `else` keyword does not seem to be needed here since the `if` block always returns `True`. So, what do you think of the slightly changed logic below?

```python
self.cpu_util_series.append(float(cpu_util_pct))
self.mem_util_series.append(float(mem_util_pct))
self.cuda_mem_util_series.append(float(cuda_mem_util_pct))
if len(self.cpu_util_series) < window_size:
    return True

self.cpu_util_series.pop(0)
self.mem_util_series.pop(0)
avg_cpu_util = sum(self.cpu_util_series) / len(self.cpu_util_series)
avg_mem_util = sum(self.mem_util_series) / len(self.mem_util_series)
cuda_mem_util = sum(self.cuda_mem_util_series) / len(self.cuda_mem_util_series)
if self.threshold_condition == "or":
    ...
```